packages/flutter/lib/src/widgets/two_dimensional_viewport.dart

```diff
@@ -0,0 +1,1328 @@
+    for (final RenderBox child in _children.values) {
+      final TwoDimensionalViewportParentData childParentData = parentDataOf(child);
+      final Rect childRect = childParentData.paintOffset & child.size;
+      if (childRect.contains(position)) {
+        result.addWithPaintOffset(
+          offset: childParentData.paintOffset,
+          position: position,
+          hitTest: (BoxHitTestResult result, Offset transformed) {
+            assert(transformed == position - childParentData.paintOffset);
+            return child.hitTest(result, position: transformed);
+          },
+        );
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /// The dimensions of the viewport.
+  ///
+  /// This [Size] represents the width and height of the visible area.
+  Size get viewportDimension {
+    assert(hasSize);
+    return size;
+  }
+
+  @override
+  void performResize() {
+    final Size? oldSize = hasSize ? size : null;
+    super.performResize();
+    // Ignoring return value since we are doing a layout either way
+    // (performLayout will be invoked next).
+    horizontalOffset.applyViewportDimension(size.width);
+    verticalOffset.applyViewportDimension(size.height);
+    if (oldSize != size) {
+      // Specs can depend on viewport size.
+      _didResize = true;
+    }
+  }
+
+  @override
+  RevealedOffset getOffsetToReveal(RenderObject target, double alignment, { Rect? rect }) {
+    // TODO(Piinks): Add this back in follow up change (ensureVisible). <BUG>
+    return const RevealedOffset(offset: 0.0, rect: Rect.zero);
+  }
+
+  /// Should be used by subclasses to invalidate any cached metrics for the
+  /// viewport.
+  ///
+  /// This is set to true when the viewport has been resized, indicating that
+  /// any cached dimensions are invalid.
+  ///
+  /// After performLayout, the value is set to false until the viewport
+  /// dimensions are changed again in [performResize].
+  ///
+  /// Subclasses are not required to use this value, but it can be used to
+  /// safely cache layout information in between layout calls.
+  bool get didResize => _didResize;
+  bool _didResize = true;
+
+  bool _needsChildRebuild = true;
+
+  /// Should be used by subclasses to invalidate any cached data from the
+  /// [delegate].
+  ///
+  /// This value is set to false after [layoutChildSequence]. If
+  /// [markNeedsLayout] is called `withDelegateRebuild` set to true, then this
+  /// value will be updated to true, signifying any cached delegate information
+  /// needs to be updated in the next call to [layoutChildSequence].
+  ///
+  /// Subclasses are not required to use this value, but it can be used to
+  /// safely cache layout information in between layout calls.
+  @protected
+  bool get needsDelegateRebuild => _needsDelegateRebuild;
+  bool _needsDelegateRebuild = true;
+
+  @override
+  void markNeedsLayout({bool withChildRebuild = false, bool withDelegateRebuild = false}) {
```

Comment:
what's the difference between `withChildRebuild` and `withDelegateRebuild`?

Maybe add a doc comment here since this method provides more functionality (due to the parameters) than the base class does and docuemnts?

Reply:
Some observations:
* `needsDelegateRebuild` appears to be unused. Should we remove it until it is useful? Or what is the intended use?
* All call sites of `markNeedsLayout` either set both arguments to true or both (implicitly) to false. Is it actually important that we differentiate between them? Should there only be one parameter instead?

Reply:
I removed needsChildRebuild. I feel a bit silly, there are obviously so many iterations on this, I really appreciate your feedback helping to clean it up.